instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/ServerInstrumenter.java

```diff
@@ -29,4 +34,133 @@ public Context start(Context parentContext, REQUEST request) {
     Context extracted = propagators.getTextMapPropagator().extract(parentContext, request, getter);
     return super.start(extracted, request);
   }
+
+  private static <REQUEST, RESPONSE> InstrumenterBuilder<REQUEST, RESPONSE> addClientIpExtractor(
+      InstrumenterBuilder<REQUEST, RESPONSE> builder, TextMapGetter<REQUEST> getter) {
+    HttpAttributesExtractor<REQUEST, RESPONSE> httpAttributesExtractor = null;
+    NetAttributesExtractor<REQUEST, RESPONSE> netAttributesExtractor = null;
+    for (AttributesExtractor<? super REQUEST, ? super RESPONSE> extractor :
+        builder.attributesExtractors) {
+      if (extractor instanceof NetAttributesExtractor) {
+        netAttributesExtractor = (NetAttributesExtractor<REQUEST, RESPONSE>) extractor;
+      } else if (extractor instanceof HttpAttributesExtractor) {
+        httpAttributesExtractor = (HttpAttributesExtractor<REQUEST, RESPONSE>) extractor;
+      }
+    }
+    if (httpAttributesExtractor == null) {
+      // Don't add HTTP_CLIENT_IP if there are no HTTP attributes registered.
+      return builder;
+    }
+    builder.addAttributesExtractor(new HttpClientIpExtractor(getter, netAttributesExtractor));
+    return builder;
+  }
+
+  private static class HttpClientIpExtractor<REQUEST, RESPONSE>
+      extends AttributesExtractor<REQUEST, RESPONSE> {
+
+    private final TextMapGetter<REQUEST> getter;
+    @Nullable private final NetAttributesExtractor<REQUEST, RESPONSE> netAttributesExtractor;
+
+    HttpClientIpExtractor(
+        TextMapGetter<REQUEST> getter,
+        @Nullable NetAttributesExtractor<REQUEST, RESPONSE> netAttributesExtractor) {
+      this.getter = getter;
+      this.netAttributesExtractor = netAttributesExtractor;
+    }
+
+    @Override
+    protected void onStart(AttributesBuilder attributes, REQUEST request) {}
+
+    @Override
+    protected void onEnd(
+        AttributesBuilder attributes, REQUEST request, @Nullable RESPONSE response) {
+      String clientIp = getForwardedClientIp(request);
+      if (clientIp == null && netAttributesExtractor != null) {
+        clientIp = netAttributesExtractor.peerIp(request, response);
```

Comment:
it's not clear to me that `http.client_ip` should be populated in this case since easy for exporter/backend to fallback to `net.peer.ip`

Reply:
We do it right now, though I agree it's not clear. Figured we'd stick to what we have for now though.

Reply:
makes sense, I'll open a spec issue to seek clarification and we can revisit later